### PR TITLE
Add comprehensive integration tests for memory sync with mocked S3

### PR DESCRIPTION
**Pull request by clusterfudge**
## Summary

This PR adds comprehensive integration tests for the memory sync engine that use a real memory proxy service with mocked S3 backend.

### New Integration Tests
- **File:** `tests/developer/test_memory_sync_integration.py`
- **Test count:** 13 integration tests
- **Status:** ✅ All passing
- **Duration:** ~1.9 seconds

### Test Coverage

Tests cover all major sync workflows:
- ✅ First sync uploads and downloads  
- ✅ Bidirectional sync
- ✅ Conflict detection when both sides modify
- ✅ Local and remote modification detection
- ✅ Operation logging for all sync operations
- ✅ State persistence across instances
- ✅ Multi-file sync (tested with 15 files)
- ✅ Nested directory structures

### Fixed Production Code

**File:** `silica/developer/memory/sync.py`

Fixed several issues to work correctly with the Memory Proxy Client API:
1. Updated `analyze_sync_operations()` to handle `SyncIndexResponse` objects
2. Updated `_determine_operation()` to use `FileMetadata` instead of dicts
3. Fixed `upload_file()` to correctly unpack tuple return values
4. Fixed `download_file()` to correctly unpack tuple return values
5. Removed invalid `expected_version=None` usage

### Test Infrastructure

- Uses real FastAPI Memory Proxy service (in-process via TestClient)
- Uses moto for accurate S3 behavior mocking
- Mocks authentication dependency for isolated testing
- No external dependencies required
- Fast execution and fully deterministic

## Test Results

```
13 passed in 1.91s
```

All pre-commit hooks passed (autoflake, ruff, ruff-format).

## Benefits

1. **Confidence** - Real end-to-end testing ensures all components work together
2. **Fast** - No external dependencies, runs in ~2 seconds
3. **Maintainable** - Clear test structure and good documentation
4. **Comprehensive** - Covers all major workflows and edge cases
5. **Deterministic** - Fully mocked, no flaky tests

### CODE REVIEW

File: silica/developer/memory/sync.py [hunk 1]
```diff
@@ -443,3 +539,654 @@ def truncate_after_sync(self, keep_days: int = 7) -> int:
...
+        Returns:
+            SyncResult with operation results
+        """
+        import time
+
+        start_time = time.time()
+        result = SyncResult()
+
+        # Skip conflicts
+        if plan.conflicts:
```
Comment (clusterfudge) [inline silica/developer/memory/sync.py]: why are we skipping conflicts? Is this intended to come in a later phase?

File: silica/developer/memory/sync.py [hunk 2]
```diff
@@ -443,3 +539,654 @@ def truncate_after_sync(self, keep_days: int = 7) -> int:
...
+
+        # Execute uploads
+        for op in plan.upload:
+            try:
+                success = self.upload_file(op.path, op.remote_version or 0)
+                if success:
+                    result.succeeded.append(op)
+                else:
+                    result.failed.append(op)
+            except Exception as e:
```
[the remark this answers is not in the document]
Comment (clusterfudge) [reply silica/developer/memory/sync.py]: and at a higher level, we should re-plan to figure out what went wrong. if the plans differ, then we should attempt the new plan. if the plans are the same, :womp:. 

consider what are retryable vs fatal errors.

**Review by clusterfudge — COMMENTED** [review]
I'm not sure that we're quite there on the sync algorithm.  from a pure user experience perspective, we should expect `retry to completion` to work for sync, with conflicts being resolved by LLM merge. We should not continue blindly when we see a conflict (ever), and should instead assume that our sync state (off the index) is out of state, and come up with a new sync plan based on that. we should be able to come up with that plan _entirely_ by comparing indexes, without having to compare the files themselves. we don't care about timestamps of file modification, we care about the version of the index at which it was modified. please review again, and target something _extremely simple_.

**Comment by clusterfudge** [comment]
Thanks for the detailed review! I understand the concerns. You're right that the current algorithm is too complex and doesn't properly handle conflicts. Let me summarize my understanding of what needs to change:

## Key Issues to Address

1. **Over-complicated logic** - Currently comparing MD5s, timestamps, and versions in convoluted ways
2. **Conflict skipping** - Conflicts are added to the plan but then skipped during execution (line 817)
3. **No LLM merge** - Conflicts should be resolved via LLM merge, not reported to user
4. **No retry/re-plan** - Should re-analyze after errors and retry

## Simplified Approach

The algorithm should be **version-based only**:

### For each file:
- If `remote.version > index.version` → remote changed
- If `local.md5 != index.md5` → local changed
- If **both changed** → **CONFLICT** (stop and resolve via LLM)
- If **only remote changed** → **DOWNLOAD**
- If **only local changed** → **UPLOAD**
- If **neither changed** → **IN SYNC**

### Conflict Resolution Flow:
1. Detect conflict → **STOP** immediately (don't continue sync)
2. Download both versions
3. Call LLM to merge
4. Upload merged result
5. **RE-PLAN** from scratch
6. Retry sync

### Error Handling:
- Version conflict (412) → Index is stale, re-plan
- Network error → Retry
- Auth error → Fatal, fail immediately
- After any error → Re-plan and try again

## Questions Before I Refactor:

1. **LLM merge interface** - Should I create a new `ConflictResolver` class with an LLM-based merge method, or integrate it into `SyncEngine` directly?

2. **Retry coordinator** - Should this be a new top-level function like `sync_with_retry()` that wraps the analyze/execute cycle?

3. **Index version tracking** - I see `index_version` in `SyncIndexResponse` - should we be using this for "has anything changed" checks rather than per-file versions?

I'll wait for your guidance on these questions before refactoring, to make sure I'm aligned with your vision.

**Comment by clusterfudge** [comment]
## Answers to Clarifying Questions

Thanks for the guidance! Here's my understanding:

### 1. LLM Merge Interface

✅ Create abstract `ConflictResolver` base class with mockable interface:
```python
class ConflictResolver(ABC):
    @abstractmethod
    def resolve_conflict(
        self, 
        path: str,
        local_content: bytes,
        remote_content: bytes,
        base_content: bytes | None  # From index if available
    ) -> bytes:
        """Returns merged content."""
        pass
```

Then implement `LLMConflictResolver` using haiku-4.5 (will check model map for full name).

### 2. Retry Coordinator

✅ Top-level function with `max_retries` kwarg:
```python
def sync_with_retry(
    sync_engine: SyncEngine,
    conflict_resolver: ConflictResolver,
    max_retries: int = 3
) -> SyncResult:
    """Retry sync until success or exhaustion."""
```

Will alert user when retries exhausted.

### 3. Version Strategy - **Critical Clarification Needed**

This is the key insight I was missing! Let me confirm I understand:

**Version Number Rules:**
- ✅ Index version = max version of any file in the index
- ✅ File version = index version when that file was last changed
- ✅ **Only remote creates versions** (monotonically increasing timestamps)
- ✅ **Local always writes to remote first**, then updates local copy/index on success

**Implications for Sync Flow:**

**Upload Flow:**
1. Local file changed (detect via MD5 vs index)
2. Call `client.write_blob(expected_version=index_entry.version)` 
3. Get back `new_version` from remote
4. Write local file (already have it)
5. Update local index with `new_version`

**Download Flow:**
1. Remote changed (detect via `remote.version > index.version`)
2. Call `client.read_blob()` to get content + version
3. Write to local file
4. Update local index with remote's version

**Question:** If local changes a file and tries to upload, but remote has ALSO changed (version conflict), we get 412. At that point:
- Do we immediately fetch remote version and trigger LLM merge?
- Or do we re-plan (which would detect the conflict) and then merge?

I assume **re-plan approach** is cleaner:
1. Upload fails with 412
2. Mark as retryable error
3. Re-analyze (will show conflict since remote.version > index.version AND local changed)
4. Resolve conflict via LLM
5. Upload merged result
6. Continue

Is this correct?

### Potential Issue I See

**Race condition during conflict resolution:**

1. We detect conflict (remote v5, index v3, local changed)
2. Download remote content
3. LLM merges local + remote
4. We try to upload with `expected_version=5`
5. But what if remote changed again to v6 while we were merging?
6. We get another 412 and have to retry

This seems fine - we'd just detect the new conflict and merge again. The retry loop handles this naturally.

**Is this the expected behavior?** Or should we have some kind of "lock" during conflict resolution?

Let me know if my understanding is correct and I'll proceed with the refactoring!

**Comment by clusterfudge** [comment]
## ⚠️ CI Failure - Old Unit Test Mocks Need Updating

The CI is failing on the old unit tests in `test_memory_sync.py` (10 failures). These failures are due to **outdated mocks** that don't match the updated API:

### Issues:

1. **`get_sync_index` mocks** - Returning `[]` instead of `SyncIndexResponse` object
2. **`write_blob` mocks** - Returning `int` instead of tuple `(is_new, md5, version)`
3. **`read_blob` mocks** - Returning `(content, metadata)` instead of tuple of 5 values

### Good News:

✅ **Integration tests: 13/13 passing** (tests with real Memory Proxy service)
✅ **Conflict resolution tests: 9/9 passing** (new functionality tests)

The integration tests exercise the real code paths and APIs, so the actual functionality is working correctly.

### Options:

**Option 1: Update the old unit tests**
- Fix the mocks to return correct types
- Update assertions to match new API
- Ensures both unit and integration tests pass

**Option 2: Remove/deprecate old unit tests**
- Integration tests provide comprehensive coverage
- Unit tests may be redundant now
- Simpler to maintain

I can quickly fix the old unit tests if you'd like, or we could discuss deprecating them since the integration tests cover the functionality more comprehensively. What's your preference?

**Comment by clusterfudge** [comment]
## ✅ Tests Fixed!

I've updated all the unit test mocks to match the new API. All tests are now passing locally:

### Test Results (Local)

```
tests/developer/test_memory_sync.py .................... 43 passed
tests/developer/test_memory_sync_integration.py ........ 13 passed
tests/developer/test_conflict_resolution.py ............ 9 passed

Total: 65/65 tests passing in 2.27s
```

### Changes Made

**Fixed mock return values:**
1. `get_sync_index` → Returns `SyncIndexResponse` object (not list)
2. `write_blob` → Returns `(is_new, md5, version)` tuple (not int)
3. `read_blob` → Returns 5-value tuple (not 2-value tuple)

**Added helper:**
- `make_sync_index_response()` - Creates proper `SyncIndexResponse` from list of dicts

CI is currently running and should pass now. All 865+ tests in the repo should pass.

**Review by clusterfudge — COMMENTED** [review]
some more notes

File: silica/developer/memory/sync_coordinator.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+    # NotFoundError is fatal (file was deleted during sync)
+    if isinstance(e, NotFoundError):
+        return True
+
+    # Default: assume fatal for safety
+    return True
+
+
+def _is_fatal_failure(op: Any) -> bool:
```
Comment (clusterfudge) [inline silica/developer/memory/sync_coordinator.py]: this method doesn't seem useful?
